REGRESSION (UI-side compositing) Disabled scrollbars don't show in some cases

nmoucht opened the pull request:
#### 5c367f97364c91adb1982fe3ede5bfd7b0fa6940
<pre>
REGRESSION (UI-side compositing) Disabled scrollbars don&apos;t show in some cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=261295">https://bugs.webkit.org/show_bug.cgi?id=261295</a>
<a href="https://rdar.apple.com/115137778">rdar://115137778</a>

Reviewed by Simon Fraser.

For the case where we need to show disabled scrollbars on a scroller without a layer we need
to have the scrollbars controller of type ScrollbarsControllerMac, so that we can create
a temporary NSScrollerImp in the web process to display the disabled scrollbars (similar to
what we do for svg foreign object, etc.). However, the previous fix checked usesAsyncScrolling
to determine this, but it is better to check usesCompositedScrolling (as for this case
usesAsyncScrolling returns true even when we don&apos;t have a layer). For the normal case we will
properly update the scrollbars controller type to RemoteScrollbarsController for when the layer
is made, and for this case where the layer isn&apos;t made, we will properly use ScrollbarsControllerMac.

* LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type-expected.txt:
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type.html:
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore:: const):
(WebCore::Internals::scrollbarsControllerTypeForNode const):
* Source/WebCore/testing/Internals.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::ensureScrollbarsController const):

Canonical link: <a href="https://commits.webkit.org/280026@main">https://commits.webkit.org/280026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dae2946cfd9b6c38a2f30a22f5a26d816b1aee6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58442 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5888 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57583 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6082 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44655 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4029 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25783 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29477 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4032 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51410 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5386 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60032 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5511 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52084 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31693 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47860 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51549 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32776 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8186 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31441 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type.html:
```diff
@@ -60,6 +60,13 @@
         </foreignObject>
     </svg>
+
+    <div style="height: 200px; width: 100px; overflow: scroll; -webkit-overflow-scrolling: touch;" id="disabledScroller">
```
`-webkit-overflow-scrolling: touch;` is a no-op on macOS (other than possibly creating a CSS stacking context).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -3325,7 +3325,7 @@ ExceptionOr<ScrollableArea*> Internals::scrollableAreaForNode(Node* node) const
 
         auto& renderBox = *element.renderBox();
-        if (!renderBox.canBeScrolledAndHasScrollableArea())
+        if (!shouldIncludeNonscrollableAreas && !renderBox.canBeScrolledAndHasScrollableArea())
```
I think we should remove the new argument and instead of `renderBox.canBeScrolledAndHasScrollableArea` just try to get the renderBox's scrollableArea.

nmoucht replied:
Fixed.

nmoucht committed:
REGRESSION (UI-side compositing) Disabled scrollbars don't show in some cases
https://bugs.webkit.org/show_bug.cgi?id=261295
rdar://115137778

Reviewed by Simon Fraser.

For the case where we need to show disabled scrollbars on a scroller without a layer we need
to have the scrollbars controller of type ScrollbarsControllerMac, so that we can create
a temporary NSScrollerImp in the web process to display the disabled scrollbars (similar to
what we do for svg foreign object, etc.). However, the previous fix checked usesAsyncScrolling
to determine this, but it is better to check usesCompositedScrolling (as for this case
usesAsyncScrolling returns true even when we don't have a layer). For the normal case we will
properly update the scrollbars controller type to RemoteScrollbarsController for when the layer
is made, and for this case where the layer isn't made, we will properly use ScrollbarsControllerMac.

* LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type-expected.txt:
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbars-controller-type.html:
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore:: const):
(WebCore::Internals::scrollbarsControllerTypeForNode const):
* Source/WebCore/testing/Internals.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::ensureScrollbarsController const):

Canonical link: https://commits.webkit.org/280026@main
